core/include/assert.h

```diff
@@ -41,10 +65,19 @@ extern const char assert_crash_message[];
  *
  * The purpose of this macro is to help programmers find bugs in their
  * programs.
+ *
+ *
+ * With @ref DEBUG_ASSERT_VERBOSE defined this will print also the file, the
+ * line and the function this assertion failed in.
+ *
+ * @see http://pubs.opengroup.org/onlinepubs/9699919799/functions/assert.html
  */
-
-#ifdef NDEBUG
-#define assert(ignore)((void) 0)
+#define assert(cond) \
+    if (!(cond)) { \
+        printf("%s:%d: %s: Assertion '%s' failed\n", RIOT_FILE_RELATIVE, \
+                __LINE__, DEBUG_FUNC, #cond); \
+        core_panic(PANIC_ASSERT_FAIL, assert_crash_message); \
+    }
```

Comment:
Could you please try if putting the implementation in common place would reduce the .text and .rodata size?

```c
typedef struct {
    const char *file, *func, *cond;
    unsigned line;
} _assert_failure_data_t;

NORETURN void _assert_failure(_assert_failure_data_t *);

#define assert(cond) \
    if (!cond) { \
        static const _assert_failure_data_t _assert_failure_data = { \
            RIOT_FILE_RELATIVE, DEBUG_FUNC, #cond, __LINE__ \
        }; \
        _assert_failure(&_assert_failure_data); \
    }
```

```c
void _assert_failure(_assert_failure_data_t *data) NORETURN {
    printf("%s:%d: %s: Assertion '%s' failed\n",
           data.file, data.line, data.func, data.cond);
    core_panic(PANIC_ASSERT_FAIL, assert_crash_message);
}
```